onnxruntime/core/graph/graph.cc

```diff
@@ -869,6 +869,12 @@ void Graph::RemoveEdge(NodeIndex src_node_index, NodeIndex dst_node_index, int s
   if (nullptr == dst_arg) {
     ORT_THROW("Invalid destination node arg slot specified when removing edge.");
   }
+
+  if (src_arg != dst_arg) {
+    // The edge ends specified by source and destination arg slot are not referring to same node arg.
+    // It means there was no edge between these two slots before.
+    ORT_THROW("Argument type mismatch when removing edge.");
+  }
   
```

Comment:
Should we instead check when adding the edge so we prevent the issue occurring?

Reply:
Unfortunately, I think we cannot do that because of the following two situations.
1) Think of a path A->B->C. You remove B and then want to add an edge A->C. We will either have to make someone manually update the input arg of C or we have to do that in AddEdge. I think it is better to do that in AddEdge (the update of node args is quite complicated and we should hide it inside the graph as much as possible).
2) Think of A->B, you remove A and then want to add an initializer that is input to B. First you should remove the edge between A and B, then delete the node A, then update the input to B (if the A-B edge removal fails, dont remove the node either; if node removal fails, dont update B's input either). Having this check in RemoveEdge forces you to respect this order, which I think is a good thing.